CMakeLists.txt

```diff
@@ -29,32 +28,90 @@ endmacro()
-    message(FATAL_ERROR "Only use WINSSL with windows")
-endif()
-
-
 include(GNUInstallDirs)
 include(FetchContent)
+include(ExternalProject)
 include(cmake/code_coverage.cmake)
 include(cmake/sanitizer.cmake)
 include(cmake/gcc_analyze.cmake)
+
+# SSL
+if(CPR_ENABLE_SSL)
+    if(NOT CPR_DETECT_SSL_BACKEND)
+        if((NOT CPR_FORCE_OPENSSL_BACKEND) AND (NOT CPR_FORCE_WINSSL_BACKEND))
+            message(FATAL_ERROR "CPR_DETECT_SSL_BACKEND disabled and no SSL backend selected manually. Either set CPR_FORCE_OPENSSL_BACKEND or CPR_FORCE_WINSSL_BACKEND to ON! To disable SSL set CPR_ENABLE_SSL to OFF.")
+        endif()
+        if((NOT WIN32) AND CPR_FORCE_WINSSL_BACKEND)
+            message(FATAL_ERROR "WinSSL is only available on Windows! Use OpenSSL instead (CPR_FORCE_OPENSSL_BACKEND).")
+        endif()
+    endif()
+
+    find_package(OpenSSL)
+    if(OPENSSL_FOUND)
+        set(OPENSSL_AVAILABLE ON CACHE INTERNAL "" FORCE)
+        message(STATUS "OpenSSL found.")
+    else()
+        set(OPENSSL_AVAILABLE OFF CACHE INTERNAL "" FORCE)
+        message(STATUS "OpenSSL not found.")
+    endif()
+
+    if(WIN32)
+        # WinSSL is allways available on Windows:
+        set(WINSSL_AVAILABLE ON CACHE INTERNAL "" FORCE)
+        message(STATUS "WinSSL found.")
+    endif()
+
+    if(CPR_DETECT_SSL_BACKEND)
+        message(STATUS "Detecting SSL backend...")
+        if(WIN32)
+            if(WINSSL_AVAILABLE)
+                message(STATUS "SSL auto detect: Using WinSSL.")
+                set(SSL_BACKEND_USED "WinSSL")
+            elseif(OPENSSL_AVAILABLE)
+                message(STATUS "SSL auto detect: Using OpenSSL.")
+                set(SSL_BACKEND_USED "OpenSSL")
+            else()
+                message(FATAL_ERROR "No valid Windows SSL backend found! Either install WinSSL or OpenSSL or disable SSL by setting CPR_ENABLE_SSL to OFF.")
+            endif()
+        else()
+            if(OPENSSL_AVAILABLE)
+                message(STATUS "SSL auto detect: Using OpenSSL.")
+                set(SSL_BACKEND_USED "OpenSSL")
+            else()
+                message(FATAL_ERROR "No valid SSL backend found! Please install OpenSSL or disable SSL by setting CPR_ENABLE_SSL to OFF.")
+            endif()
+        endif()
+    else()
+        if(CPR_FORCE_OPENSSL_BACKEND)
+            if(OPENSSL_AVAILABLE)
+                message(STATUS "Using OpenSSL.")
+                set(SSL_BACKEND_USED "OpenSSL")
+            else()
+                message(FATAL_ERROR "CPR_FORCE_OPENSSL_BACKEND enabled but we were not able to find OpenSSL!")
+            endif()
+        elseif(CPR_FORCE_WINSSL_BACKEND)
+            if(WINSSL_AVAILABLE)
+                message(STATUS "Using WinSSL.")
+                set(SSL_BACKEND_USED "WinSSL")
+            else()
+                message(FATAL_ERROR "CPR_FORCE_OPENSSL_BACKEND enabled but we were not able to find OpenSSL!")
+            endif()
+        endif()
+    endif()
+endif()
```

Comment:
Is this massive detection logic necessary?
Can't we just rely on setting the requested configuration according to the user request?

Reply:
A project I'm working on had to switch to this branch a few days ago because OpenSSL wasn't working.

This piece of the cmakelist really threw me for a spin. I think having both `CPR_DETECT_SSL_BACKEND` and `CPR_FORCE_OPENSSL_BACKEND` as separate vars can initially be confusing.

I'd personally suggest allowing users to set a var that will directly affect `SSL_BACKEND_USED` as I initially admittedly tried to set the backend by doing `set(SSL_BACKEND_USED "OpenSSL")`

Reply:
Also it hurts to say, I think you guys are right.
I would suggest removing the `CPR_DETECT_SSL_BACKEND` variable and fall back to it in case the user did not set either `CPR_FORCE_OPENSSL_BACKEND` or `CPR_FORCE_WINSSL_BACKEND`.
This allows new users to get up and running quickly without having to think about the backend and the complexity is reduced for advanced users.
Any thoughts?

Reply:
I think that would be the right move. `CPR_FORCE_OPENSSL_BACKEND` and `CPR_FORCE_WINSSL_BACKEND` only ever do anything if `CPR_DETECT_SSL_BACKEND` is set to `OFF`.

I see no intentional reason to have `CPR_DETECT_SSL_BACKEND` `ON` while either `CPR_FORCE_OPENSSL_BACKEND` or `CPR_FORCE_WINSSL_BACKEND` are `ON`.

The `find_package` for OpenSSL should probably also only be in the `CPR_FORCE_OPENSSL_BACKEND` block as I think that will actually count as an error causing configure to be considered incomplete if not found. Which would be an issue if someone intends to use WinSSL and doesn't have OpenSSL available.
Some people will also have custom `find<PackageName>.cmake` scripts and those will be triggered by using `find_package`

Reply:
Done.